sycl/plugins/esimd_cpu/pi_esimd_cpu.cpp

```diff
@@ -593,367 +588,525 @@ pi_result piContextCreate(const pi_context_properties *Properties,
                                             const void *PrivateInfo, size_t CB,
                                             void *UserData),
                           void *UserData, pi_context *RetContext) {
-  (void)Properties;
-  (void)NumDevices;
-  (void)Devices;
-  (void)PFnNotify;
-  (void)UserData;
-  (void)RetContext;
-  DIE_NO_IMPLEMENTATION;
+  if (NumDevices != 1) {
+    return PI_INVALID_VALUE;
+  }
+  if (Devices == nullptr) {
+    return PI_INVALID_DEVICE;
+  }
+  if (RetContext == nullptr) {
+    return PI_INVALID_VALUE;
+  }
+
+  cm_support::CmDevice *device = nullptr;
+  unsigned int version = 0;
+
+  int result = cm_support::CreateCmDevice(device, version);
+
+  if (result != cm_support::CM_SUCCESS) {
+    return PI_INVALID_VALUE;
+  }
+
+  try {
+    *RetContext = new _pi_context(*Devices, device);
+  } catch (const std::bad_alloc &) {
+    return PI_OUT_OF_HOST_MEMORY;
+  } catch (...) {
+    return PI_ERROR_UNKNOWN;
+  }
   return PI_SUCCESS;
 }
 
 pi_result piContextGetInfo(pi_context, pi_context_info, size_t, void *,
                            size_t *) {
   DIE_NO_IMPLEMENTATION;
-  return PI_SUCCESS;
 }
 
 pi_result piextContextSetExtendedDeleter(pi_context,
                                          pi_context_extended_deleter, void *) {
   DIE_NO_IMPLEMENTATION;
-  return PI_SUCCESS;
 }
 
 pi_result piextContextGetNativeHandle(pi_context, pi_native_handle *) {
   DIE_NO_IMPLEMENTATION;
-  return PI_SUCCESS;
 }
 
 pi_result piextContextCreateWithNativeHandle(pi_native_handle, pi_uint32,
                                              const pi_device *, bool,
                                              pi_context *) {
   DIE_NO_IMPLEMENTATION;
-  return PI_SUCCESS;
 }
 
 pi_result piContextRetain(pi_context) {
   DIE_NO_IMPLEMENTATION;
-  return PI_SUCCESS;
 }
 
 pi_result piContextRelease(pi_context Context) {
-  (void)Context;
-  DIE_NO_IMPLEMENTATION;
+  if ((Context == nullptr) || (Context->CmDevicePtr == nullptr)) {
+    return PI_INVALID_CONTEXT;
+  }
+
+  int result = cm_support::DestroyCmDevice(Context->CmDevicePtr);
```

Comment:
It seems confusing that the device is destoroyed in the function dedicated for context releasing. Could you please clarify why it is not done in `piDeviceCreate`/`piDeviceRelease`?

Reply:
There is no `piDeviceCreate`. As far as I remember, in CM, there is one-on-one correspondence between context and device therefore `_pi_context` is defined as follows (pi_esimd_cpu.hpp)
```
struct _pi_context : _pi_object {
  _pi_context(pi_device DeviceArg, cm_support::CmDevice *CmDeviceArg)
      : Device{DeviceArg}, CmDevicePtr{CmDeviceArg} {}

  /// One-to-one mapping between Context and Device
  pi_device Device;

  cm_support::CmDevice *CmDevicePtr = nullptr;

  /// Map SVM memory starting address to corresponding
  /// CmBufferSVM object. CmBufferSVM object is needed to release memory.
  std::unordered_map<void *, cm_support::CmBufferSVM *> Addr2CmBufferSVM;
};
```
I'll unify CM device initialization and release to context implementation and put comments.

Reply:
>I'll unify CM device initialization and release to device PI_APIs.

If you don't mind, I would like to put TODO comments here about one-on-on correspondence and revise this device/context later. I remember that there was reference implementation for device/context in CM_EMU/SHIM layer implementation that I used in initial implementation for pi_esimd_cpu, but cannot find it now.

Reply:
Device management is moved out of context management and now done through piDevice* PI_APIs - `piDevicesGet`,`piDeviceRetain`, and `piDeviceRelease`